src/framework/pybind_json.hpp

```diff
@@ -257,38 +257,46 @@ json_t JSON::numpy_to_json(py::array_t<T, py::array::c_style> arr) {
 }
 
 void std::to_json(json_t &js, const py::handle &obj) {
+    auto type_str = std::string(py::str(obj.get_type()));
     if (py::isinstance<py::bool_>(obj)) {
         js = obj.cast<nl::json::boolean_t>();
         //js = obj.cast<bool>();
-    } else if (py::isinstance<py::int_>(obj)) {
-        js = obj.cast<nl::json::number_integer_t>();
-    } else if (py::isinstance<py::float_>(obj)) {
-        js = obj.cast<nl::json::number_float_t>();
     } else if (py::isinstance<py::str>(obj)) {
         js = obj.cast<nl::json::string_t>();
     } else if (py::isinstance<py::tuple>(obj) || py::isinstance<py::list>(obj)) {
         js = nl::json::array();
-        for (py::handle value: obj)
-        {
+        for (py::handle value: obj) {
             js.push_back(value);
         }
     } else if (py::isinstance<py::dict>(obj)) {
-        for (auto item : py::cast<py::dict>(obj))
-        {
+        for (auto item : py::cast<py::dict>(obj)) {
             js[item.first.cast<nl::json::string_t>()] = item.second;
         }
     } else if (py::isinstance<py::array_t<double> >(obj)) {
         js = JSON::numpy_to_json(obj.cast<py::array_t<double, py::array::c_style> >());
     } else if (py::isinstance<py::array_t<std::complex<double> > >(obj)) {
         js = JSON::numpy_to_json(obj.cast<py::array_t<std::complex<double>, py::array::c_style> >());
-    } else if (std::string(py::str(obj.get_type())) == "<class \'complex\'>") {
+    } else if (py::isinstance<py::float_>(obj)
+               || type_str == "<class \'numpy.float32\'>"
```

Comment:
You could separate the string check comparisons to all be done at the end (and sort them at the end in the order they are most likely to appear). This will prevent performance degradation for the most used types. Eg.

```cpp
else if (py::isinstance<py::float_>(obj)) {
  js = obj.cast<nl::json::number_float_t>();
} else if (py::isinstance<py::int_>(obj) {
  js = obj.cast<nl::json::number_integer_t>();
} else if (obj.is_none()) {
 return;
} else {
  // For performance reasons all string-match conversion cases are considered last
  auto type_str = std::string(py::str(obj.get_type()));
  if (std::string(py::str(obj.get_type())) == "<class \'complex\'>"
               || type_str == "<class \'numpy.complex128\'>"
               || type_str == "<class \'numpy.complex64\'>"
               || type_str == "<class \'numpy.complex_\'>" ) {
      auto tmp = obj.cast<std::complex<double>>();
      js.push_back(tmp.real());
      js.push_back(tmp.imag());
    } else if (type_str == "<class \'numpy.float64\'>" || type_str == "<class \'numpy.float32\'>") {
        js = obj.cast<nl::json::number_float_t>();
  } else if (type_str == "<class \'numpy.int64\'>"
               || type_str == "<class \'numpy.uint64\'>"
               || type_str == "<class \'numpy.int32\'>"
               || type_str == "<class \'numpy.uint32\'>" ) {
        js = obj.cast<nl::json::number_integer_t>();
    } else {
        throw std::runtime_error("to_json not implemented for this type of object: " + obj.cast<std::string>());
        throw std::runtime_error("to_json not implemented for this type of object: " + std::string(py::str(obj.get_type())));
    }
}
```

Reply:
h/t done